### ISSUE TITLE
Primary font ignores the fontconfig's charset property.

### ISSUE THREAD

**Issue by huyuehui** [alacritty/crossfont#30]
I reported [an issue](https://github.com/alacritty/alacritty/issues/5351) several days ago. Maybe at that time I didn't make myself clear. I guess that the following source blocks cause this issue.

In src/ft/mod.rs :
In my case, the function load_face_with_glyph(...) has never been called. Because in face_for_glyph(...), it always returns early inside the 'if let' block.

In the get_face(...) function, the 'coverage' of the 'Fallbacklist' struct doesn't include the charset of the primary font. So setting the charset property of the first matched font (the primary font) in fonts.conf doesn't work if no fallback is happened.

**Comment by kchibisov** [comment]
@huyuehui would you mind providing example fontconfig configuration? I'm not into its syntax wrt charsets so it could help.

**Comment by huyuehui** [comment]
@kchibisov
It is something like the following:
...
<match target="scan">
	  <test name="family"><string>Migu 1M</string></test>
	  <edit name="charset">
		  <charset>
			  
			  <range><int>0x0020</int><int>0x007E</int></range>
			  
			  <range><int>0x00A0</int><int>0x0181</int></range>
			  <int>0x0186</int><int>0x0189</int><int>0x018A</int>
			  <range><int>0x018E</int><int>0x0194</int></range>
			  <range><int>0x0198</int><int>0x019B</int></range>
			  <range><int>0x019D</int><int>0x01A1</int></range>
			  <range><int>0x01AF</int><int>0x01B1</int></range>
			  <range><int>0x01B3</int><int>0x01B6</int></range>
			  <range><int>0x01BF</int><int>0x01E3</int></range>
			  <range><int>0x01E6</int><int>0x01ED</int></range>
			  <range><int>0x01EF</int><int>0x01F5</int></range>
			  <range><int>0x01F7</int><int>0x01FF</int></range>
			  <range><int>0x0218</int><int>0x021F</int></range>
			  <range><int>0x0230</int><int>0x0233</int></range>
			  <int>0x0237</int><int>0x023D</int>
			  
			  <range><int>0x0250</int><int>0x02FF</int></range>
			  
			  <range><int>0x0372</int><int>0x0375</int></range>
			  <int>0x037A</int><int>0x037E</int><int>0x038C</int>
			  <range><int>0x0384</int><int>0x038A</int></range>
			  <range><int>0x038E</int><int>0x03A1</int></range>
			  <range><int>0x03A3</int><int>0x03E1</int></range>
			  
			  <range><int>0x0400</int><int>0x045F</int></range>
			  <int>0x0462</int><int>0x0463</int>
			  <range><int>0x0472</int><int>0x0475</int></range>
			  <int>0x0480</int><int>0x0481</int><int>0x0490</int><int>0x0491</int>
			  <int>0x04AE</int><int>0x04AF</int><int>0x04E8</int><int>0x04E9</int>
			  
			  <int>0x20A1</int><range><int>0x20A8</int><int>0x20AC</int></range>
			  <range><int>0x20AF</int><int>0x20B3</int></range><int>0x20B5</int>
			  
			  <int>0x2103</int><int>0x2113</int><int>0x2116</int><int>0x211E</int>
			  <int>0x2121</int><int>0x2122</int><int>0x2126</int><int>0x2127</int>
			  
			  <range><int>0x2160</int><int>0x216B</int></range>
			  <range><int>0x2170</int><int>0x217B</int></range>
			  
			  <range><int>0x2190</int><int>0x2199</int></range>
			  <range><int>0x21B0</int><int>0x21B5</int></range>
			  <range><int>0x21C4</int><int>0x21CA</int></range>
			  <range><int>0x21D0</int><int>0x21D9</int></range>
			  <range><int>0x21E4</int><int>0x21ED</int></range>
			  
			  <range><int>0x2400</int><int>0x2426</int></range>
			  
			  <range><int>0x2460</int><int>0x2473</int></range>
			  <range><int>0x24D0</int><int>0x24FF</int></range>
			  <range><int>0x2776</int><int>0x2793</int></range>
			  
			  <int>0x2A2F</int><int>0x2C71</int><int>0x2C7D</int>
			  
			  <range><int>0x3041</int><int>0x3096</int></range>
			  <range><int>0x3099</int><int>0x30FF</int></range>
			  <range><int>0x31F0</int><int>0x31FF</int></range>
			  
			  <int>0xA78B</int><int>0xA78C</int>
			  
			  <range><int>0xFB00</int><int>0xFB04</int></range>
			  
			  <int>0xFEFF</int>
		  </charset>
	  </edit>
  </match>
...

**Comment by kchibisov** [comment]
Ok, I'll try to take a look when I have time...

**Comment by korkota** [comment]
It would be nice to have the ability to use the fontconfig's charset property.
For example, I’d like to hide the monochrome warning icon in the JetBrains Mono font and use the colorful one from the Noto Color Emoji font.
<details>
<summary>fonts.conf example</summary>

```xml
<?xml version='1.0'?>
<!DOCTYPE fontconfig SYSTEM 'fonts.dtd'>
<fontconfig>
  <alias>
    <family>terminal-font</family>
    <prefer>
      <family>JetBrainsMono Nerd Font</family>
      <family>Noto Color Emoji</family>
    </prefer>
  </alias>
  <match target="scan">
      <test name="family">
          <string>JetBrainsMono Nerd Font</string>
      </test>
      <edit name="charset" mode="assign">
          <minus>
              <name>charset</name>
              <charset>
                  <range>
                      <int>0x26A0</int>
                      <int>0x26A1</int>
                  </range>
              </charset>
          </minus>
      </edit>
  </match>
</fontconfig>
```

</details>


<img width="1027" height="595" alt="Image" src="https://github.com/user-attachments/assets/ecc811f2-9160-44e5-ac75-4d0c629535aa" />

**Comment by chrisduerr** [comment]
@korkota This isn't a solution to this issue by any means, but with pure emoji fonts you might be able to get what you want by just picking the emoji font of your choosing first, then have your 'primary' font as fallback in fontconfig. Since emoji fonts usually only have emojis, that generally works pretty well for people that enjoy colored emojis.

**Comment by korkota** [comment]
@chrisduerr Yes, I tried to do that, but I got this funny giant caret symbol:

<img width="1282" height="798" alt="Image" src="https://github.com/user-attachments/assets/d4327947-1c49-4afb-85d3-3b89c31549d8" />

It looks like the Noto Color Emoji font can't function as a primary font.
